HandmadeMath.h

```diff
@@ -1465,6 +1516,124 @@ HMM_QuaternionV4(hmm_vec4 Vector)
+{
+    hmm_quaternion Result = {0};
+
+    Result.X = QuaternionOne.X * QuaternionTwo.W + QuaternionOne.Y * QuaternionTwo.Z - QuaternionOne.Z * QuaternionTwo.Y + QuaternionOne.W * QuaternionTwo.X;
+    Result.Y = -QuaternionOne.X * QuaternionTwo.Z + QuaternionOne.Y * QuaternionTwo.W + QuaternionOne.Z * QuaternionTwo.X + QuaternionOne.W * QuaternionTwo.Y;
+    Result.Z = QuaternionOne.X * QuaternionTwo.Y - QuaternionOne.Y * QuaternionTwo.X + QuaternionOne.Z * QuaternionTwo.W + QuaternionOne.W * QuaternionTwo.Z;
+    Result.W = -QuaternionOne.X * QuaternionTwo.X - QuaternionOne.Y * QuaternionTwo.Y - QuaternionOne.Z * QuaternionTwo.Z + QuaternionOne.W * QuaternionTwo.W;
+
+    return(Result);
+}
+
+HINLINE hmm_quaternion
+HMM_Quaternion_Multiply_Float(hmm_quaternion QuaternionOne, float Multiplicative)
+{
+    hmm_quaternion Result = {0};
+
+    Result.X = QuaternionOne.X * Multiplicative;
+    Result.Y = QuaternionOne.Y * Multiplicative;
+    Result.Z = QuaternionOne.Z * Multiplicative;
+    Result.W = QuaternionOne.W * Multiplicative;
+
+    return(Result);
+}
+
+HINLINE hmm_quaternion
+HMM_Quaternion_Divide(hmm_quaternion QuaternionOne, hmm_quaternion QuaternionTwo)
+{
+    hmm_quaternion Result = {0};
+
+    Result.X = QuaternionOne.X * (1/QuaternionTwo.X);
+    Result.Y = QuaternionOne.Y * (1/QuaternionTwo.Y);
+    Result.Z = QuaternionOne.Z * (1/QuaternionTwo.Z);
+    Result.W = QuaternionOne.W * (1/QuaternionTwo.W);
+
+    return(Result);
+}
+
+HINLINE float
+HMM_Quaternion_Dot(hmm_quaternion QuaternionOne, hmm_quaternion QuaternionTwo)
+{
+    float Result = 0.00f;
+
+    Result = (QuaternionOne.X * QuaternionTwo.X) + (QuaternionOne.Y * QuaternionTwo.Y) + (QuaternionOne.Z * QuaternionTwo.Z) + (QuaternionOne.W * QuaternionTwo.W);
+
+    return(Result);
+}
+
+HINLINE hmm_quaternion
+HMM_Quaternion_Slerp(hmm_quaternion QuaternionOne, hmm_quaternion QuaternionTwo, float Time)
+{
+    hmm_quaternion Result = {0};
+    hmm_quaternion QuaternionLeft = {0};
+    hmm_quaternion QuaternionRight = {0};
+
+    float Cos_Theta = HMM_Quaternion_Dot(QuaternionOne, QuaternionTwo);
+    float Angle = HMM_ACosF(Cos_Theta);
+    
+    float S1 = HMM_SinF(1.0f - Time*Angle);
+    float S2 = HMM_SinF(Time*Angle);
+    float Is = 1.0f / HMM_SinF(Angle);
+
+   QuaternionLeft.X = QuaternionTwo.X * S1;
+   QuaternionLeft.Y = QuaternionTwo.Y * S1;
+   QuaternionLeft.Z = QuaternionTwo.Z * S1;
+   QuaternionLeft.W = QuaternionTwo.W * S1;
+
+   QuaternionRight.X = QuaternionTwo.X * S2;
+   QuaternionRight.Y = QuaternionTwo.Y * S2;
+   QuaternionRight.Z = QuaternionTwo.Z * S2;
+   QuaternionRight.W = QuaternionTwo.W * S2;
+
+   Result.X = QuaternionLeft.X + QuaternionRight.X;
+   Result.Y = QuaternionLeft.Y + QuaternionRight.Y;
+   Result.Z = QuaternionLeft.Z + QuaternionRight.Z;
+   Result.W = QuaternionLeft.W + QuaternionRight.W;
+
+   Result.X = Result.X * Is;
+   Result.Y = Result.Y * Is;
+   Result.Z = Result.Z * Is;
+   Result.W = Result.W * Is;
```

Comment:
@StrangeZak Do you know if it would be better or worse for performance to put everything into one line? Like, for example:

```cpp
Result.X = (QuaternionOne.X * S1 + QuaternionTwo.X * S2) * Is;
```

Certainly we could avoid defining two extra quaternions inside this function, but for all I know that's better for performance anyway...